### ISSUE TITLE
Semantic Highlighting for Type Hints

### ISSUE THREAD

**Issue by hassanselim0** [microsoft/pylance-release#335]
I recently updated to Pylance 2020.9.0 and I really like the Semantic Highlighting!

However it seems like type hints don't have any special semantic (usually it's a `class`) and I'm wanted type hints to have a different color than classes. Whether it's annotation type hints or comment type hints.

Before this update, comment type hints had special TextMate scopes like `comment.typehint` and `meta.typehint` which I've customized, and now these are overridden by the semantic colors. Is it at least possible to still enforce my colors in this case?

**Comment by hassanselim0** [comment]
This is a screenshot of how comment type hints look right now after the update:
![image](https://user-images.githubusercontent.com/1431770/92425320-2ad98980-f187-11ea-871b-6b7d9b2fde4e.png)
Previously the classes (`int`, `Order`, `datetime`) had the same color as the `# type:` token (a semi-transparent purple), is it possible to achieve that again without completely disabling semantic highlighting?

**Comment by erictraut** [comment]
Out of curiosity, why are you still using type comments? Are you stuck on an old version of Python? Since Python 3.6, variable type annotations can and should be provided using the new syntax in [PEP 526](https://www.python.org/dev/peps/pep-0526/).

**Comment by hassanselim0** [comment]
I used the type comments mainly because I was able to give them a different color based on TextMate Scopes. Not being able to customize the color for annotations made Django Model Definitions quite unreadable. I use annotations everywhere else though.

**Comment by huguesv** [comment]
Thanks. I'll investigate if we can make this customizable.

**Comment by huguesv** [comment]
This has been fixed for the next release.

![image](https://user-images.githubusercontent.com/1696845/93515183-0bf09b00-f8dd-11ea-8606-b7fd7b6c2eed.png)

**Comment by huguesv** [comment]
You'll also be able to do something like this:
![image](https://user-images.githubusercontent.com/1696845/93515764-faf45980-f8dd-11ea-92ce-83310bc522da.png)

Given settings like these:
```
  "editor.semanticTokenColorCustomizations": {
    "[One Dark Pro]": {
      "rules": {
        "*.typeHint": "#00ff00"
      }
    }
  },
  "editor.tokenColorCustomizations": {
    "[One Dark Pro]": {
      "textMateRules": [
        {
          "scope": "meta.typehint",
          "settings": {
            "foreground": "#a32f9aaa"
          }
        },
        {
          "scope": "comment.typehint",
          "settings": {
            "foreground": "#a32f9aaa"
          }
        }
      ]
    }
  },
```

**Comment by hassanselim0** [comment]
@huguesv great 😁
Does modifier also apply to annotation-based type hints?
```python
self._identifier: int = MY_CONSTANT
```

**Comment by huguesv** [comment]
@hassanselim0 it does not, from what I saw it didn't seem that textmate was setting a special scope for it either.

![image](https://user-images.githubusercontent.com/1696845/93624878-e1aee400-f995-11ea-9539-b4c2872a4bdf.png)

But it's relatively easy to add a different modifier for this case (so that you'd be able to distinguish from the ones in comments).

**Comment by hassanselim0** [comment]
@huguesv It would be great to have a modifier for annotation type hints (same or different), the reason I went with comment hints in the first place was because TestMate didn't have a scope for annotation hints.
I think if I can give annotation hints a different color it would make certain bits of code more readable, like the Django Model in the screenshot I uploaded earlier, to visually separate the type from the field class beside it. That's my main goal here.

Side Note: The reason I'm even putting type hints in Django Models is because even though the class attributes have types that are a subclass of orm `Field`, the instance attributes have normal python types. So `name = CharField()` when later accessed via `self.name` gives an `str` type. So I write the class attr as `name: str = CharField()` or `name = CharField()  # type: str` so that VS Code would better help me.
If there is a way to write a `pyi` file and define `CharField` in such a way that the the Language Server would automatically show me all instances of `CharField` as having a type of `str` without using any type hints, I'd be glad to know.

**Comment by erictraut** [comment]
@hassanselim0, the best way to declare `CharField` in a stub file is to define a `__get__` method that returns `str`. That's how `CharField` is implemented in django, so defining it that way in the type stub will produce the most accurate results.

Incidentally, there is a stub package available for django. That might save you a bunch of time. https://pypi.org/project/django-stubs/

**Comment by jakebailey** [comment]
We ship those stubs with pylance, so I would have expected for this to have worked (if the stub was defined in this way).

**Comment by hassanselim0** [comment]
I do have django stubs installed and working, but it doesn't do that behavior in my code.
I checked to see the stub file here: https://github.com/typeddjango/django-stubs/blob/master/django-stubs/db/models/fields/__init__.pyi#L107
They have overloaded `__get__`, it can return a Field type or a "get return type", but I don't understand how the type checker is supposed to figure out which overload to pick. The end result is that Pylance tells me the instance attr is of type `Any`. I'm not sure if this is a bug in the stub or in Pylance, then maybe create a separate issue for this?

**Comment by jakebailey** [comment]
This would warrant another issue, yes, as it doesn't really fit within the bounds of the original semantic highlighting issue.

**Comment by erictraut** [comment]
Yeah, if you think that the type checker is not correctly handling `__get__` in this case, please file a separate issue — preferably in the pyright repo, since that's where the type checking logic lives. Thanks!

**Comment by hassanselim0** [comment]
Done, I've actually commented on a closed ticket about a very similar issue, so it can be reopened in case it's a regression.

Getting back to the original issue, I still think it would be useful to have a semantic modifier for annotation type hints, so you can easily customize the colors for all type hints whether they are in an annotation or a comment.

**Comment by huguesv** [comment]
I've just checked in a change that allows customization of colors for type hints in annotation or comment. It will be in the next release of Pylance.

Default colors (One Dark Pro):
![image](https://user-images.githubusercontent.com/1696845/93636519-ecbf3f80-f9a8-11ea-9e42-f1e36ff9d305.png)


Customized colors (One Dark Pro):
```
  "editor.semanticTokenColorCustomizations": {
    "[One Dark Pro]": {
      "rules": {
        "*.typeHint": "#286605",
        "*.typeHintComment": "#6ed3d3"
      }
    }
  }
```

![image](https://user-images.githubusercontent.com/1696845/93636564-fcd71f00-f9a8-11ea-9cd3-f0b11dcdf5a5.png)

**Comment by hassanselim0** [comment]
@huguesv thank you 😁

**Comment by jakebailey** [comment]
This issue has been fixed in version 2020.9.6, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/master/CHANGELOG.md#202096-23-september-2020

**Comment by hassanselim0** [comment]
@jakebailey @huguesv Thank you for your efforts, I just got around to testing this today and stumbled upon two issues. I set `*.typeHint` and `*.typeHintComment` to `#ff000000` to test the new modifiers.

This is the first one:
![image](https://user-images.githubusercontent.com/1431770/95119271-ffc56400-074b-11eb-94ca-106d94d49d24.png)
The square brackets used for generic types doesn't have the new `typeHint` modifier.

This is the second one:
![image](https://user-images.githubusercontent.com/1431770/95119535-664a8200-074c-11eb-960d-cbcf38850ab7.png)
In function argument annotations that have generic types, the generic type doesn't have the `typeHint` modifier.
